src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -538,26 +536,22 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
 		 *         is established between the control websocket and the cloud service
 		 */
 		public CompletableFuture<Void> openAsync(Duration timeout) {
-			// Establish a WebSocket connection right now so we can detect any fatal errors
-			return CompletableFuture.runAsync(() -> {
-				CompletableFuture<Void> connectTask = null;
-				boolean success = false;
-				
-				try {
-					// Block so we surface any errors to the user right away.
-					connectTask = this.ensureConnectTask(timeout);
-					connectTask.join();
-
+			// Establish a WebSocket connection right now so we can detect any fatal errors		
+			CompletableFuture<ClientWebSocket> connectTask = this.ensureConnectTask(timeout);
+			return connectTask.thenApply((webSocket) -> {
 					this.tokenRenewer.setOnTokenRenewed((token) -> this.onTokenRenewed(token));
-					this.receivePumpAsync();
-					success = true;
-				} finally {
-					if (!success) {
-						this.closeOrAbortWebSocketAsync(connectTask, new CloseReason(CloseCodes.UNEXPECTED_CONDITION,
-								"closing web socket connection because something went wrong trying to connect."));			
+					this.receivePumpAsync(); // This starts pumping but doesn't wait
+					return (Void)null;
+				}).whenComplete(($void, err) -> {
+					if (err != null) {
+						// catch (Throwable err) {
+						RelayLogger.throwingException(err, this.listener);
+						CloseReason closeReason = new CloseReason(CloseCodes.UNEXPECTED_CONDITION,
+								"closing web socket connection because something went wrong trying to connect.");
```

Comment:
>web socket [](start = 17, length = 10)

Do you think saying "control web socket" would be more specific?

Reply:
Sure, I don't think I actually touched the value of this String.

---
In reply to: [258630907](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258630907) [](ancestors = 258630907)